{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5baf8e96",
   "metadata": {},
   "outputs": [],
   "source": [
    "# standard data analysis\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "import os\n",
    "\n",
    "# convenience\n",
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ede1d1d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_csv_to_wig_file(input_file, output_file, track_name, track_description, \n",
    "                          value_column='TISCounts', transform=None, \n",
    "                          input_dirpath='/lab/barcheese01/smaffa/filtered_tis_data',\n",
    "                          output_dirpath='/lab/barcheese01/smaffa/igv_files'):\n",
    "    if '.csv' in input_file:\n",
    "        tis_df = pd.read_csv(os.path.join(input_dirpath, input_file))\n",
    "    elif '.txt' in input_file:\n",
    "        tis_df = pd.read_csv(os.path.join(input_dirpath, input_file), sep='\\t')\n",
    "\n",
    "    # trim the input table into the necessary columns for the output wig file\n",
    "    df_slim = tis_df.loc[:, [value_column, 'GenomePos']]\n",
    "    genome_annotations = df_slim['GenomePos'].str.split(':', expand=True)\n",
    "    genome_annotations.columns = ['chr', 'span', 'strand']\n",
    "    spans = genome_annotations['span'].str.split('-', expand=True)\n",
    "    spans.columns = ['start', 'end']\n",
    "    df_slim = pd.concat([df_slim, genome_annotations, spans], axis=1)\n",
    "    df_slim['locus'] = df_slim.apply(lambda x: x['start'] if x['strand'] == '+' else x['end'] if x['strand'] == '-' else 'None', axis=1)\n",
    "\n",
    "    if transform is not None:\n",
    "        df_slim[value_column] = transform(df_slim[value_column])\n",
    "\n",
    "    with open(os.path.join(output_dirpath, output_file), \"w\") as f:\n",
    "        for chr, idx in df_slim.groupby('chr').groups.items():\n",
    "            if 'chr' in chr:\n",
    "                f.write(f'track type=wiggle_0 name={track_name} description={track_description}\\n')\n",
    "                f.write(f'variableStep chrom={chr} span=1\\n')\n",
    "                subset = df_slim.loc[idx, :].sort_values('locus')\n",
    "                for locus, counts in zip(subset['locus'].tolist(), subset[value_column].tolist()):\n",
    "                    if counts != 0:\n",
    "                        f.write(f'{locus} {counts}\\n')\n",
    "    print(f'Saved file to {os.path.join(output_dirpath, output_file)}')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a2a0c26",
   "metadata": {},
   "source": [
    "# Global wig files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0b3465a4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved file to /lab/barcheese01/smaffa/igv_files/HeLa_all_ribotish_counts.wig\n"
     ]
    }
   ],
   "source": [
    "write_csv_to_wig_file(\n",
    "    input_file='HeLa_TIS_predict_all.txt',\n",
    "    output_file='HeLa_all_ribotish_counts.wig',\n",
    "    track_name = 'HeLa_all_ribotish_counts',\n",
    "    track_description = 'All ribotish TIS counts',\n",
    "    input_dirpath='/lab/barcheese01/aTIS_data/ribotish'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "963bc6c3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/lab/barcheese01/smaffa/miniforge3/envs/riboseq/lib/python3.11/site-packages/pandas/core/arraylike.py:402: RuntimeWarning: divide by zero encountered in log10\n",
      "  result = getattr(ufunc, method)(*inputs, **kwargs)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved file to /lab/barcheese01/smaffa/igv_files/HeLa_all_ribotish_qvals.wig\n"
     ]
    }
   ],
   "source": [
    "write_csv_to_wig_file(\n",
    "    input_file='HeLa_TIS_predict_all.txt',\n",
    "    output_file='HeLa_all_ribotish_qvals.wig',\n",
    "    value_column='TISQvalue',\n",
    "    transform=lambda x: (-np.log10(x)).replace([np.inf, -np.inf], 100).fillna(0),\n",
    "    track_name = 'HeLa_all_ribotish_log10_qval',\n",
    "    track_description = 'All ribotish TISDiff q-values',\n",
    "    input_dirpath='/lab/barcheese01/aTIS_data/ribotish'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "1f8a291c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved file to /lab/barcheese01/smaffa/igv_files/HeLa_all_ribotish_frameq.wig\n"
     ]
    }
   ],
   "source": [
    "write_csv_to_wig_file(\n",
    "    input_file='HeLa_TIS_predict_all.txt',\n",
    "    output_file='HeLa_all_ribotish_frameq.wig',\n",
    "    value_column='FrameQvalue',\n",
    "    transform=lambda x: (-np.log10(x)).replace([np.inf, -np.inf], 100).fillna(0),\n",
    "    track_name = 'HeLa_all_ribotish_log10_frame_qval',\n",
    "    track_description = 'All ribotish TISDiff q-values',\n",
    "    input_dirpath='/lab/barcheese01/aTIS_data/ribotish'\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c05ef15",
   "metadata": {},
   "source": [
    "# Strategy-specific wig files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "88cd2756",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved file to /lab/barcheese01/smaffa/igv_files/HeLa_exclusion_window_30nt.wig\n"
     ]
    }
   ],
   "source": [
    "write_csv_to_wig_file(\n",
    "    input_file='HeLa_TIS_filtered.csv',\n",
    "    output_file='HeLa_exclusion_window_30nt.wig',\n",
    "    track_name = 'HeLa_exclusion_window_30nt',\n",
    "    track_description = 'Filtering by iteratively highest readcount, then exclusion of TISs within 30nt downstream'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "adef4384",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved file to /lab/barcheese01/smaffa/igv_files/HeLa_exclusion_window_30nt_mincount_5.wig\n"
     ]
    }
   ],
   "source": [
    "write_csv_to_wig_file(\n",
    "    input_file='HeLa_TIS_filtered2.csv',\n",
    "    output_file='HeLa_exclusion_window_30nt_mincount_5.wig',\n",
    "    track_name = 'HeLa_exclusion_window_30nt_mincount_5',\n",
    "    track_description = 'Filtering by iteratively highest readcount, then exclusion of TISs within 30nt downstream, with a minimum readcount of 5'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "bbf1af63",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved file to /lab/barcheese01/smaffa/igv_files/HeLa_exclusion_window_45nt.wig\n"
     ]
    }
   ],
   "source": [
    "write_csv_to_wig_file(\n",
    "    input_file='HeLa_TIS_filtered3.csv',\n",
    "    output_file='HeLa_exclusion_window_45nt.wig',\n",
    "    track_name = 'HeLa_exclusion_window_45nt',\n",
    "    track_description = 'Filtering by iteratively highest readcount, then exclusion of TISs within 45nt downstream'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "9eea3a93",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved file to /lab/barcheese01/smaffa/igv_files/HeLa_peak_finding_qval_increase_order2_mincount_5.wig\n"
     ]
    }
   ],
   "source": [
    "write_csv_to_wig_file(\n",
    "    input_file='HeLa_TIS_alt_filtered.csv',\n",
    "    output_file='HeLa_peak_finding_qval_increase_order2_mincount_5.wig',\n",
    "    track_name = 'HeLa_peak_finding_qval_increase_order2_mincount_5',\n",
    "    track_description = 'Filtering by peak finding using a minimum readcount of 5 and a TISDiff increase in significance by 10^2'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "52df9f95",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved file to /lab/barcheese01/smaffa/igv_files/HeLa_peak_finding_qval_increase_order3_mincount_5.wig\n"
     ]
    }
   ],
   "source": [
    "write_csv_to_wig_file(\n",
    "    input_file='HeLa_TIS_alt_filtered2.csv',\n",
    "    output_file='HeLa_peak_finding_qval_increase_order3_mincount_5.wig',\n",
    "    track_name = 'HeLa_peak_finding_qval_increase_order3_mincount_5',\n",
    "    track_description = 'Filtering by peak finding using a minimum readcount of 5 and a TISDiff increase in significance by 10^3'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13a8a7d3",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'pd' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mNameError\u001b[39m                                 Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[1]\u001b[39m\u001b[32m, line 1\u001b[39m\n\u001b[32m----> \u001b[39m\u001b[32m1\u001b[39m tis_df = \u001b[43mpd\u001b[49m.read_csv(os.path.join(\u001b[33m'\u001b[39m\u001b[33m/lab/barcheese01/smaffa/filtered_tis_data\u001b[39m\u001b[33m'\u001b[39m, \u001b[33m'\u001b[39m\u001b[33mHeLa_TIS_filtered.csv\u001b[39m\u001b[33m'\u001b[39m))\n\u001b[32m      2\u001b[39m alt_tis_df = pd.read_csv(os.path.join(\u001b[33m'\u001b[39m\u001b[33m/lab/barcheese01/smaffa/filtered_tis_data\u001b[39m\u001b[33m'\u001b[39m, \u001b[33m'\u001b[39m\u001b[33mHeLa_TIS_alt_filtered.csv\u001b[39m\u001b[33m'\u001b[39m))\n",
      "\u001b[31mNameError\u001b[39m: name 'pd' is not defined"
     ]
    }
   ],
   "source": [
    "tis_df = pd.read_csv(os.path.join('/lab/barcheese01/smaffa/filtered_tis_data', 'HeLa_TIS_filtered.csv'))\n",
    "alt_tis_df = pd.read_csv(os.path.join('/lab/barcheese01/smaffa/filtered_tis_data', 'HeLa_TIS_alt_filtered.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "6c5bef44",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(24963, 19)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_csv(os.path.join('/lab/barcheese01/smaffa/filtered_tis_data', 'HeLa_TIS_filtered.csv')).shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "51d8538c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(17298, 19)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_csv(os.path.join('/lab/barcheese01/smaffa/filtered_tis_data', 'HeLa_TIS_filtered2.csv')).shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "d26ab7f1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(15842, 19)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_csv(os.path.join('/lab/barcheese01/smaffa/filtered_tis_data', 'HeLa_TIS_filtered3.csv')).shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2d8d26c",
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mValueError\u001b[39m                                Traceback (most recent call last)",
      "\u001b[32m/tmp/ipykernel_3131743/1926787358.py\u001b[39m in \u001b[36m?\u001b[39m\u001b[34m()\u001b[39m\n\u001b[32m----> \u001b[39m\u001b[32m1\u001b[39m pd.read_csv(os.path.join('/lab/barcheese01/smaffa/filtered_tis_data', 'HeLa_TIS_alt_filtered.csv')).assign(\n\u001b[32m      2\u001b[39m     RecatTisType=\u001b[38;5;28;01mlambda\u001b[39;00m x:\n\u001b[32m      3\u001b[39m         x[\u001b[33m'TisType'\u001b[39m] \u001b[38;5;28;01mif\u001b[39;00m x[\u001b[33m'TisType'\u001b[39m] == \u001b[33m'0'\u001b[39m \u001b[38;5;28;01melse\u001b[39;00m\n\u001b[32m      4\u001b[39m         \u001b[33m'Annotated'\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m (\u001b[33m'Annotated'\u001b[39m \u001b[38;5;28;01min\u001b[39;00m x[\u001b[33m'TisType'\u001b[39m]) \u001b[38;5;28;01mor\u001b[39;00m (\u001b[33m'Known'\u001b[39m \u001b[38;5;28;01min\u001b[39;00m x[\u001b[33m'TisType'\u001b[39m]) \u001b[38;5;28;01melse\u001b[39;00m\n",
      "\u001b[32m/lab/barcheese01/smaffa/miniforge3/envs/riboseq/lib/python3.11/site-packages/pandas/core/frame.py\u001b[39m in \u001b[36m?\u001b[39m\u001b[34m(self, **kwargs)\u001b[39m\n\u001b[32m   5711\u001b[39m         \"\"\"\n\u001b[32m   5712\u001b[39m         data = self.copy(deep=\u001b[38;5;28;01mFalse\u001b[39;00m)\n\u001b[32m   5713\u001b[39m \n\u001b[32m   5714\u001b[39m         \u001b[38;5;28;01mfor\u001b[39;00m k, v \u001b[38;5;28;01min\u001b[39;00m kwargs.items():\n\u001b[32m-> \u001b[39m\u001b[32m5715\u001b[39m             data[k] = com.apply_if_callable(v, data)\n\u001b[32m   5716\u001b[39m         \u001b[38;5;28;01mreturn\u001b[39;00m data\n",
      "\u001b[32m/lab/barcheese01/smaffa/miniforge3/envs/riboseq/lib/python3.11/site-packages/pandas/core/common.py\u001b[39m in \u001b[36m?\u001b[39m\u001b[34m(maybe_callable, obj, **kwargs)\u001b[39m\n\u001b[32m    387\u001b[39m     \"\"\"\n\u001b[32m    388\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m isinstance(maybe_callable, Expression):\n\u001b[32m    389\u001b[39m         \u001b[38;5;28;01mreturn\u001b[39;00m maybe_callable._eval_expression(obj, **kwargs)\n\u001b[32m    390\u001b[39m     \u001b[38;5;28;01melif\u001b[39;00m callable(maybe_callable):\n\u001b[32m--> \u001b[39m\u001b[32m391\u001b[39m         \u001b[38;5;28;01mreturn\u001b[39;00m maybe_callable(obj, **kwargs)\n\u001b[32m    392\u001b[39m \n\u001b[32m    393\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m maybe_callable\n",
      "\u001b[32m/tmp/ipykernel_3131743/1926787358.py\u001b[39m in \u001b[36m?\u001b[39m\u001b[34m(x)\u001b[39m\n\u001b[32m      2\u001b[39m     RecatTisType=\u001b[38;5;28;01mlambda\u001b[39;00m x:\n",
      "\u001b[32m/lab/barcheese01/smaffa/miniforge3/envs/riboseq/lib/python3.11/site-packages/pandas/core/generic.py\u001b[39m in \u001b[36m?\u001b[39m\u001b[34m(self)\u001b[39m\n\u001b[32m   1511\u001b[39m     @final\n\u001b[32m   1512\u001b[39m     \u001b[38;5;28;01mdef\u001b[39;00m __bool__(self) -> NoReturn:\n\u001b[32m-> \u001b[39m\u001b[32m1513\u001b[39m         raise ValueError(\n\u001b[32m   1514\u001b[39m             \u001b[33mf\"The truth value of a {type(self).__name__} is ambiguous. \"\u001b[39m\n\u001b[32m   1515\u001b[39m             \u001b[33m\"Use a.empty, a.bool(), a.item(), a.any() or a.all().\"\u001b[39m\n\u001b[32m   1516\u001b[39m         )\n",
      "\u001b[31mValueError\u001b[39m: The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all()."
     ]
    }
   ],
   "source": [
    "pd.read_csv(os.path.join('/lab/barcheese01/smaffa/filtered_tis_data', 'HeLa_TIS_alt_filtered.csv')).assign(\n",
    "    RecatTisType=lambda x: \n",
    "        \n",
    "        'Annotated' if ('Annotated' in x['TisType']) or ('Known' in x['TisType']) else\n",
    "        'Truncated' if 'Truncated' in x['TisType'] else\n",
    "        'uORF' if \"5' UTR\" in x['TisType'] else\n",
    "        'Extended' if 'Extended' in x['TisType'] else\n",
    "        'Other',\n",
    "        axis=1\n",
    ").value_counts(['RecatTisType'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6e3b4cb2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "TisType\n",
       "3'UTR                          2\n",
       "3'UTR:CDSFrameOverlap          1\n",
       "3'UTR:Known                    2\n",
       "5'UTR                        379\n",
       "5'UTR:CDSFrameOverlap          7\n",
       "5'UTR:Known                    8\n",
       "Annotated                   6414\n",
       "Extended                     309\n",
       "Extended:CDSFrameOverlap      12\n",
       "Extended:Known                 9\n",
       "Internal                      16\n",
       "Internal:CDSFrameOverlap       2\n",
       "Internal:Known                 1\n",
       "Truncated                    664\n",
       "Truncated:Known               32\n",
       "Name: Symbol, dtype: int64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_csv(os.path.join('/lab/barcheese01/smaffa/filtered_tis_data', 'HeLa_TIS_alt_filtered.csv')).groupby(['TisType']).nunique()['Symbol']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8b659340",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "TisType\n",
       "3'UTR                          9\n",
       "3'UTR:CDSFrameOverlap        509\n",
       "3'UTR:Known                   71\n",
       "5'UTR                        447\n",
       "5'UTR:CDSFrameOverlap         93\n",
       "5'UTR:Known                   77\n",
       "Annotated                   5956\n",
       "Extended                    1825\n",
       "Extended:CDSFrameOverlap     330\n",
       "Extended:Known                98\n",
       "Internal                      65\n",
       "Internal:CDSFrameOverlap     154\n",
       "Internal:Known                25\n",
       "Novel                        329\n",
       "Novel:CDSFrameOverlap       2100\n",
       "Novel:Known                 1558\n",
       "Truncated                   5029\n",
       "Truncated:Known              508\n",
       "Name: Symbol, dtype: int64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_csv(os.path.join('/lab/barcheese01/aTIS_data/ribotish/HeLa_TIS_filtered_predictor.txt'), sep='\\t').groupby(['TisType']).nunique()['Symbol']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "64341f24",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(10114, 21)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_csv(os.path.join('/lab/barcheese01/smaffa/filtered_tis_data', 'HeLa_TIS_alt_filtered2.csv')).shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2d20544d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Symbol\n",
       "AHNAK     35\n",
       "MAP4      30\n",
       "KYNU      28\n",
       "POLR1A    27\n",
       "ITPR3     26\n",
       "Name: GenomePos, dtype: int64"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tis_df.groupby('Symbol').nunique()['GenomePos'].sort_values(ascending=False).head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4e403a32",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Symbol\n",
       "LDHA    14\n",
       "TFG     12\n",
       "CLU     12\n",
       "TMX2    11\n",
       "ASPH    10\n",
       "Name: GenomePos, dtype: int64"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alt_tis_df.groupby('Symbol').nunique()['GenomePos'].sort_values(ascending=False).head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "fabe4191",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gid</th>\n",
       "      <th>Tid</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>StartCodon</th>\n",
       "      <th>Start</th>\n",
       "      <th>TisType</th>\n",
       "      <th>TISGroup</th>\n",
       "      <th>TISCounts</th>\n",
       "      <th>TISQvalue</th>\n",
       "      <th>GenomePos</th>\n",
       "      <th>Seq</th>\n",
       "      <th>AASeq</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>MANE_Select</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16854</th>\n",
       "      <td>ENSG00000047849.23</td>\n",
       "      <td>ENST00000395734.8</td>\n",
       "      <td>MAP4</td>\n",
       "      <td>AAG</td>\n",
       "      <td>2648</td>\n",
       "      <td>Truncated</td>\n",
       "      <td>5</td>\n",
       "      <td>21</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>chr3:47852791-47871239:-</td>\n",
       "      <td>AAGTCTGTACCAGCTGACTTGAGTCGCCCAAAGAGCACCTCCACCA...</td>\n",
       "      <td>KSVPADLSRPKSTSTSSMKKTTTLSGTAPAAGVVPSRVKATPMPSR...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16850</th>\n",
       "      <td>ENSG00000047849.23</td>\n",
       "      <td>ENST00000395734.8</td>\n",
       "      <td>MAP4</td>\n",
       "      <td>AAG</td>\n",
       "      <td>2498</td>\n",
       "      <td>Truncated</td>\n",
       "      <td>5</td>\n",
       "      <td>32</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>chr3:47852791-47872019:-</td>\n",
       "      <td>AAGAGCACTCAGACTGTTGCAAAAACCACAACAGCTGCTGCTGTTG...</td>\n",
       "      <td>KSTQTVAKTTTAAAVASTGPSSRSPSTLLPKKPTAIKTEGKPAEVK...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16853</th>\n",
       "      <td>ENSG00000047849.23</td>\n",
       "      <td>ENST00000395734.8</td>\n",
       "      <td>MAP4</td>\n",
       "      <td>AAG</td>\n",
       "      <td>2225</td>\n",
       "      <td>Truncated</td>\n",
       "      <td>5</td>\n",
       "      <td>21</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>chr3:47852791-47875876:-</td>\n",
       "      <td>AAGACTTCAACATCGAAAGCCAAAACACAGCCCACTTCTCTCCCTA...</td>\n",
       "      <td>KTSTSKAKTQPTSLPKQPAPTTIGGLNKKPMSLASGLVPAAPPKRP...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16852</th>\n",
       "      <td>ENSG00000047849.23</td>\n",
       "      <td>ENST00000395734.8</td>\n",
       "      <td>MAP4</td>\n",
       "      <td>GTG</td>\n",
       "      <td>1802</td>\n",
       "      <td>Truncated</td>\n",
       "      <td>5</td>\n",
       "      <td>23</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>chr3:47852791-47916119:-</td>\n",
       "      <td>GTGACTCCAGCCAAAGATGTTCCACCACTCTCAGAAACAGAGGCAA...</td>\n",
       "      <td>VTPAKDVPPLSETEATPVPIKDMEIAQTQKGISEDSHLESLQDVGQ...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16851</th>\n",
       "      <td>ENSG00000047849.23</td>\n",
       "      <td>ENST00000395734.8</td>\n",
       "      <td>MAP4</td>\n",
       "      <td>AAG</td>\n",
       "      <td>1316</td>\n",
       "      <td>Truncated</td>\n",
       "      <td>5</td>\n",
       "      <td>23</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>chr3:47852791-47916605:-</td>\n",
       "      <td>AAGGATTTGGTATTACTCTCAGAAATAGAGGTGGCACAGGCTAATG...</td>\n",
       "      <td>KDLVLLSEIEVAQANDIISSTEISSAEKVALSSETEVALARDMTLP...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16849</th>\n",
       "      <td>ENSG00000047849.23</td>\n",
       "      <td>ENST00000395734.8</td>\n",
       "      <td>MAP4</td>\n",
       "      <td>ATG</td>\n",
       "      <td>95</td>\n",
       "      <td>Annotated</td>\n",
       "      <td>5</td>\n",
       "      <td>91</td>\n",
       "      <td>1.346053e-02</td>\n",
       "      <td>chr3:47852791-47998860:-</td>\n",
       "      <td>ATGGCTGACCTCAGTCTTGCAGATGCATTAACAGAACCATCTCCAG...</td>\n",
       "      <td>MADLSLADALTEPSPDIEGEIKRDFIATLEAEAFDDVVGETVGKTD...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24685</th>\n",
       "      <td>ENSG00000047849.23</td>\n",
       "      <td>ENST00000683076.1</td>\n",
       "      <td>MAP4</td>\n",
       "      <td>AAG</td>\n",
       "      <td>6726</td>\n",
       "      <td>Truncated</td>\n",
       "      <td>4</td>\n",
       "      <td>17</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>chr3:47852933-47855319:-</td>\n",
       "      <td>AAGGAGAAGGCCCAGGCCAAGGTGGGATCCCTCGATAATGTGGGCC...</td>\n",
       "      <td>KEKAQAKVGSLDNVGHLPAGGAVKTEGGGSEAPLCPGPPAGEEPAI...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24678</th>\n",
       "      <td>ENSG00000047849.23</td>\n",
       "      <td>ENST00000683076.1</td>\n",
       "      <td>MAP4</td>\n",
       "      <td>AAG</td>\n",
       "      <td>6567</td>\n",
       "      <td>Truncated</td>\n",
       "      <td>4</td>\n",
       "      <td>37</td>\n",
       "      <td>2.341482e-01</td>\n",
       "      <td>chr3:47852933-47867281:-</td>\n",
       "      <td>AAGGACAATATTAAGCATGTCCCTGGAGGTGGTAATGTTCAGATTC...</td>\n",
       "      <td>KDNIKHVPGGGNVQIQNKKVDISKVSSKCGSKANIKHKPGGGDVKI...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24687</th>\n",
       "      <td>ENSG00000047849.23</td>\n",
       "      <td>ENST00000683076.1</td>\n",
       "      <td>MAP4</td>\n",
       "      <td>AAG</td>\n",
       "      <td>6555</td>\n",
       "      <td>Truncated</td>\n",
       "      <td>4</td>\n",
       "      <td>13</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>chr3:47852933-47867293:-</td>\n",
       "      <td>AAGTGTGGTTCCAAGGACAATATTAAGCATGTCCCTGGAGGTGGTA...</td>\n",
       "      <td>KCGSKDNIKHVPGGGNVQIQNKKVDISKVSSKCGSKANIKHKPGGG...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24682</th>\n",
       "      <td>ENSG00000047849.23</td>\n",
       "      <td>ENST00000683076.1</td>\n",
       "      <td>MAP4</td>\n",
       "      <td>AAG</td>\n",
       "      <td>6090</td>\n",
       "      <td>Truncated</td>\n",
       "      <td>4</td>\n",
       "      <td>21</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>chr3:47852933-47871239:-</td>\n",
       "      <td>AAGTCTGTACCAGCTGACTTGAGTCGCCCAAAGAGCACCTCCACCA...</td>\n",
       "      <td>KSVPADLSRPKSTSTSSMKKTTTLSGTAPAAGVVPSRVKATPMPSR...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24679</th>\n",
       "      <td>ENSG00000047849.23</td>\n",
       "      <td>ENST00000683076.1</td>\n",
       "      <td>MAP4</td>\n",
       "      <td>AAG</td>\n",
       "      <td>5940</td>\n",
       "      <td>Truncated</td>\n",
       "      <td>4</td>\n",
       "      <td>32</td>\n",
       "      <td>5.004593e-01</td>\n",
       "      <td>chr3:47852933-47872019:-</td>\n",
       "      <td>AAGAGCACTCAGACTGTTGCAAAAACCACAACAGCTGCTGCTGTTG...</td>\n",
       "      <td>KSTQTVAKTTTAAAVASTGPSSRSPSTLLPKKPTAIKTEGKPAEVK...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24683</th>\n",
       "      <td>ENSG00000047849.23</td>\n",
       "      <td>ENST00000683076.1</td>\n",
       "      <td>MAP4</td>\n",
       "      <td>AAG</td>\n",
       "      <td>5667</td>\n",
       "      <td>Truncated</td>\n",
       "      <td>4</td>\n",
       "      <td>21</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>chr3:47852933-47875876:-</td>\n",
       "      <td>AAGACTTCAACATCGAAAGCCAAAACACAGCCCACTTCTCTCCCTA...</td>\n",
       "      <td>KTSTSKAKTQPTSLPKQPAPTTIGGLNKKPMSLASGLVPAAPPKRP...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24681</th>\n",
       "      <td>ENSG00000047849.23</td>\n",
       "      <td>ENST00000683076.1</td>\n",
       "      <td>MAP4</td>\n",
       "      <td>GTG</td>\n",
       "      <td>1809</td>\n",
       "      <td>Truncated</td>\n",
       "      <td>4</td>\n",
       "      <td>23</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>chr3:47852933-47916119:-</td>\n",
       "      <td>GTGACTCCAGCCAAAGATGTTCCACCACTCTCAGAAACAGAGGCAA...</td>\n",
       "      <td>VTPAKDVPPLSETEATPVPIKDMEIAQTQKGISEDSHLESLQDVGQ...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24688</th>\n",
       "      <td>ENSG00000047849.23</td>\n",
       "      <td>ENST00000683076.1</td>\n",
       "      <td>MAP4</td>\n",
       "      <td>AAG</td>\n",
       "      <td>1485</td>\n",
       "      <td>Truncated</td>\n",
       "      <td>4</td>\n",
       "      <td>12</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>chr3:47852933-47916443:-</td>\n",
       "      <td>AAGGATAAAGCACTACCTTTAGAAGCAGAGGTGGCCCCAGTCAAGG...</td>\n",
       "      <td>KDKALPLEAEVAPVKDMAQLPETEIAPAKDVAPSTVKEVGLLKDMS...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24680</th>\n",
       "      <td>ENSG00000047849.23</td>\n",
       "      <td>ENST00000683076.1</td>\n",
       "      <td>MAP4</td>\n",
       "      <td>AAG</td>\n",
       "      <td>1323</td>\n",
       "      <td>Truncated</td>\n",
       "      <td>4</td>\n",
       "      <td>23</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>chr3:47852933-47916605:-</td>\n",
       "      <td>AAGGATTTGGTATTACTCTCAGAAATAGAGGTGGCACAGGCTAATG...</td>\n",
       "      <td>KDLVLLSEIEVAQANDIISSTEISSAEKVALSSETEVALARDMTLP...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24684</th>\n",
       "      <td>ENSG00000047849.23</td>\n",
       "      <td>ENST00000683076.1</td>\n",
       "      <td>MAP4</td>\n",
       "      <td>AAG</td>\n",
       "      <td>885</td>\n",
       "      <td>Truncated</td>\n",
       "      <td>4</td>\n",
       "      <td>18</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>chr3:47852933-47917043:-</td>\n",
       "      <td>AAGGACATGGCACTAGCTACAAAAACCGAGGTGGCATTGGCTAAAG...</td>\n",
       "      <td>KDMALATKTEVALAKDMESPTKLDVTLAKDMQPSMESDMALVKDME...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24686</th>\n",
       "      <td>ENSG00000047849.23</td>\n",
       "      <td>ENST00000683076.1</td>\n",
       "      <td>MAP4</td>\n",
       "      <td>AAG</td>\n",
       "      <td>771</td>\n",
       "      <td>Truncated</td>\n",
       "      <td>4</td>\n",
       "      <td>14</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>chr3:47852933-47917157:-</td>\n",
       "      <td>AAGGAGATAGAAATGGCATCAGAAGAGAGGCCACCAGCACAAGCAT...</td>\n",
       "      <td>KEIEMASEERPPAQALEIMMGLKTTDMAPSKETEMALAKDMALATK...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24677</th>\n",
       "      <td>ENSG00000047849.23</td>\n",
       "      <td>ENST00000683076.1</td>\n",
       "      <td>MAP4</td>\n",
       "      <td>ATG</td>\n",
       "      <td>102</td>\n",
       "      <td>Annotated</td>\n",
       "      <td>4</td>\n",
       "      <td>91</td>\n",
       "      <td>3.792186e-05</td>\n",
       "      <td>chr3:47852933-47998860:-</td>\n",
       "      <td>ATGGCTGACCTCAGTCTTGCAGATGCATTAACAGAACCATCTCCAG...</td>\n",
       "      <td>MADLSLADALTEPSPDIEGEIKRDFIATLEAEAFDDVVGETVGKTD...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11652</th>\n",
       "      <td>ENSG00000047849.23</td>\n",
       "      <td>ENST00000360240.10</td>\n",
       "      <td>MAP4</td>\n",
       "      <td>AAG</td>\n",
       "      <td>3549</td>\n",
       "      <td>Truncated</td>\n",
       "      <td>5</td>\n",
       "      <td>37</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>chr3:47853031-47867281:-</td>\n",
       "      <td>AAGGACAATATTAAGCATGTCCCTGGAGGTGGTAATGTTCAGATTC...</td>\n",
       "      <td>KDNIKHVPGGGNVQIQNKKVDISKVSSKCGSKANIKHKPGGGDVKI...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11657</th>\n",
       "      <td>ENSG00000047849.23</td>\n",
       "      <td>ENST00000360240.10</td>\n",
       "      <td>MAP4</td>\n",
       "      <td>AAG</td>\n",
       "      <td>3072</td>\n",
       "      <td>Truncated</td>\n",
       "      <td>5</td>\n",
       "      <td>21</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>chr3:47853031-47871239:-</td>\n",
       "      <td>AAGTCTGTACCAGCTGACTTGAGTCGCCCAAAGAGCACCTCCACCA...</td>\n",
       "      <td>KSVPADLSRPKSTSTSSMKKTTTLSGTAPAAGVVPSRVKATPMPSR...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11653</th>\n",
       "      <td>ENSG00000047849.23</td>\n",
       "      <td>ENST00000360240.10</td>\n",
       "      <td>MAP4</td>\n",
       "      <td>AAG</td>\n",
       "      <td>2922</td>\n",
       "      <td>Truncated</td>\n",
       "      <td>5</td>\n",
       "      <td>32</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>chr3:47853031-47872019:-</td>\n",
       "      <td>AAGAGCACTCAGACTGTTGCAAAAACCACAACAGCTGCTGCTGTTG...</td>\n",
       "      <td>KSTQTVAKTTTAAAVASTGPSSRSPSTLLPKKPTAIKTEGKPAEVK...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11656</th>\n",
       "      <td>ENSG00000047849.23</td>\n",
       "      <td>ENST00000360240.10</td>\n",
       "      <td>MAP4</td>\n",
       "      <td>AAG</td>\n",
       "      <td>2649</td>\n",
       "      <td>Truncated</td>\n",
       "      <td>5</td>\n",
       "      <td>21</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>chr3:47853031-47875876:-</td>\n",
       "      <td>AAGACTTCAACATCGAAAGCCAAAACACAGCCCACTTCTCTCCCTA...</td>\n",
       "      <td>KTSTSKAKTQPTSLPKQPAPTTIGGLNKKPMSLASGLVPAAPPKRP...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11654</th>\n",
       "      <td>ENSG00000047849.23</td>\n",
       "      <td>ENST00000360240.10</td>\n",
       "      <td>MAP4</td>\n",
       "      <td>GTG</td>\n",
       "      <td>2226</td>\n",
       "      <td>Truncated</td>\n",
       "      <td>5</td>\n",
       "      <td>23</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>chr3:47853031-47916119:-</td>\n",
       "      <td>GTGACTCCAGCCAAAGATGTTCCACCACTCTCAGAAACAGAGGCAA...</td>\n",
       "      <td>VTPAKDVPPLSETEATPVPIKDMEIAQTQKGISEDSHLESLQDVGQ...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11655</th>\n",
       "      <td>ENSG00000047849.23</td>\n",
       "      <td>ENST00000360240.10</td>\n",
       "      <td>MAP4</td>\n",
       "      <td>AAG</td>\n",
       "      <td>1740</td>\n",
       "      <td>Truncated</td>\n",
       "      <td>5</td>\n",
       "      <td>23</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>chr3:47853031-47916605:-</td>\n",
       "      <td>AAGGATTTGGTATTACTCTCAGAAATAGAGGTGGCACAGGCTAATG...</td>\n",
       "      <td>KDLVLLSEIEVAQANDIISSTEISSAEKVALSSETEVALARDMTLP...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11650</th>\n",
       "      <td>ENSG00000047849.23</td>\n",
       "      <td>ENST00000360240.10</td>\n",
       "      <td>MAP4</td>\n",
       "      <td>ATG</td>\n",
       "      <td>519</td>\n",
       "      <td>Annotated</td>\n",
       "      <td>5</td>\n",
       "      <td>91</td>\n",
       "      <td>1.346053e-02</td>\n",
       "      <td>chr3:47853031-47998860:-</td>\n",
       "      <td>ATGGCTGACCTCAGTCTTGCAGATGCATTAACAGAACCATCTCCAG...</td>\n",
       "      <td>MADLSLADALTEPSPDIEGEIKRDFIATLEAEAFDDVVGETVGKTD...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11651</th>\n",
       "      <td>ENSG00000047849.23</td>\n",
       "      <td>ENST00000360240.10</td>\n",
       "      <td>MAP4</td>\n",
       "      <td>CTG</td>\n",
       "      <td>402</td>\n",
       "      <td>Extended</td>\n",
       "      <td>5</td>\n",
       "      <td>42</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>chr3:47853031-48088870:-</td>\n",
       "      <td>CTGGCTCCCGGCTCCTCCCGACGTCTCCTACCTCCTCACGGCTCTT...</td>\n",
       "      <td>LAPGSSRRLLPPHGSSRRSPGSLLPQLRLGGGGQLQWCRMADLSLA...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18224</th>\n",
       "      <td>ENSG00000047849.23</td>\n",
       "      <td>ENST00000423088.5</td>\n",
       "      <td>MAP4</td>\n",
       "      <td>ATG</td>\n",
       "      <td>48</td>\n",
       "      <td>Annotated</td>\n",
       "      <td>6</td>\n",
       "      <td>91</td>\n",
       "      <td>2.265140e-01</td>\n",
       "      <td>chr3:47916863-47998860:-</td>\n",
       "      <td>ATGGCTGACCTCAGTCTTGCAGATGCATTAACAGAACCATCTCCAG...</td>\n",
       "      <td>MADLSLADALTEPSPDIEGEIKRDFIATLEAEAFDDVVGETVGKTD...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>3.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18615</th>\n",
       "      <td>ENSG00000047849.23</td>\n",
       "      <td>ENST00000434267.5</td>\n",
       "      <td>MAP4</td>\n",
       "      <td>AAG</td>\n",
       "      <td>161</td>\n",
       "      <td>Truncated</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>chr3:47975457-47998797:-</td>\n",
       "      <td>AAGCGGGACTTCATTGCCACACTAGAGGCAGAGGCCTTTGATGATG...</td>\n",
       "      <td>KRDFIATLEAEAFDDVVGETVGKTDYIPLLDVDEKTGNSESKKKPC...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18824</th>\n",
       "      <td>ENSG00000047849.23</td>\n",
       "      <td>ENST00000439356.2</td>\n",
       "      <td>MAP4</td>\n",
       "      <td>AAG</td>\n",
       "      <td>134</td>\n",
       "      <td>Truncated</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>chr3:47975457-47998797:-</td>\n",
       "      <td>AAGCGGGACTTCATTGCCACACTAGAGGCAGAGGCCTTTGATGATG...</td>\n",
       "      <td>KRDFIATLEAEAFDDVVGETVGKTDYIPLLDVDEKTGNSESKKKPC...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18616</th>\n",
       "      <td>ENSG00000047849.23</td>\n",
       "      <td>ENST00000434267.5</td>\n",
       "      <td>MAP4</td>\n",
       "      <td>ATA</td>\n",
       "      <td>158</td>\n",
       "      <td>Truncated</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>chr3:47975457-47998800:-</td>\n",
       "      <td>ATAAAGCGGGACTTCATTGCCACACTAGAGGCAGAGGCCTTTGATG...</td>\n",
       "      <td>IKRDFIATLEAEAFDDVVGETVGKTDYIPLLDVDEKTGNSESKKKP...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18823</th>\n",
       "      <td>ENSG00000047849.23</td>\n",
       "      <td>ENST00000439356.2</td>\n",
       "      <td>MAP4</td>\n",
       "      <td>ATG</td>\n",
       "      <td>71</td>\n",
       "      <td>Annotated</td>\n",
       "      <td>3</td>\n",
       "      <td>91</td>\n",
       "      <td>2.367954e-08</td>\n",
       "      <td>chr3:47975457-47998860:-</td>\n",
       "      <td>ATGGCTGACCTCAGTCTTGCAGATGCATTAACAGAACCATCTCCAG...</td>\n",
       "      <td>MADLSLADALTEPSPDIEGEIKRDFIATLEAEAFDDVVGETVGKTD...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18614</th>\n",
       "      <td>ENSG00000047849.23</td>\n",
       "      <td>ENST00000434267.5</td>\n",
       "      <td>MAP4</td>\n",
       "      <td>ATG</td>\n",
       "      <td>98</td>\n",
       "      <td>Annotated</td>\n",
       "      <td>2</td>\n",
       "      <td>91</td>\n",
       "      <td>1.650892e-12</td>\n",
       "      <td>chr3:47975457-47998860:-</td>\n",
       "      <td>ATGGCTGACCTCAGTCTTGCAGATGCATTAACAGAACCATCTCCAG...</td>\n",
       "      <td>MADLSLADALTEPSPDIEGEIKRDFIATLEAEAFDDVVGETVGKTD...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Gid                 Tid Symbol StartCodon  Start  \\\n",
       "16854  ENSG00000047849.23   ENST00000395734.8   MAP4        AAG   2648   \n",
       "16850  ENSG00000047849.23   ENST00000395734.8   MAP4        AAG   2498   \n",
       "16853  ENSG00000047849.23   ENST00000395734.8   MAP4        AAG   2225   \n",
       "16852  ENSG00000047849.23   ENST00000395734.8   MAP4        GTG   1802   \n",
       "16851  ENSG00000047849.23   ENST00000395734.8   MAP4        AAG   1316   \n",
       "16849  ENSG00000047849.23   ENST00000395734.8   MAP4        ATG     95   \n",
       "24685  ENSG00000047849.23   ENST00000683076.1   MAP4        AAG   6726   \n",
       "24678  ENSG00000047849.23   ENST00000683076.1   MAP4        AAG   6567   \n",
       "24687  ENSG00000047849.23   ENST00000683076.1   MAP4        AAG   6555   \n",
       "24682  ENSG00000047849.23   ENST00000683076.1   MAP4        AAG   6090   \n",
       "24679  ENSG00000047849.23   ENST00000683076.1   MAP4        AAG   5940   \n",
       "24683  ENSG00000047849.23   ENST00000683076.1   MAP4        AAG   5667   \n",
       "24681  ENSG00000047849.23   ENST00000683076.1   MAP4        GTG   1809   \n",
       "24688  ENSG00000047849.23   ENST00000683076.1   MAP4        AAG   1485   \n",
       "24680  ENSG00000047849.23   ENST00000683076.1   MAP4        AAG   1323   \n",
       "24684  ENSG00000047849.23   ENST00000683076.1   MAP4        AAG    885   \n",
       "24686  ENSG00000047849.23   ENST00000683076.1   MAP4        AAG    771   \n",
       "24677  ENSG00000047849.23   ENST00000683076.1   MAP4        ATG    102   \n",
       "11652  ENSG00000047849.23  ENST00000360240.10   MAP4        AAG   3549   \n",
       "11657  ENSG00000047849.23  ENST00000360240.10   MAP4        AAG   3072   \n",
       "11653  ENSG00000047849.23  ENST00000360240.10   MAP4        AAG   2922   \n",
       "11656  ENSG00000047849.23  ENST00000360240.10   MAP4        AAG   2649   \n",
       "11654  ENSG00000047849.23  ENST00000360240.10   MAP4        GTG   2226   \n",
       "11655  ENSG00000047849.23  ENST00000360240.10   MAP4        AAG   1740   \n",
       "11650  ENSG00000047849.23  ENST00000360240.10   MAP4        ATG    519   \n",
       "11651  ENSG00000047849.23  ENST00000360240.10   MAP4        CTG    402   \n",
       "18224  ENSG00000047849.23   ENST00000423088.5   MAP4        ATG     48   \n",
       "18615  ENSG00000047849.23   ENST00000434267.5   MAP4        AAG    161   \n",
       "18824  ENSG00000047849.23   ENST00000439356.2   MAP4        AAG    134   \n",
       "18616  ENSG00000047849.23   ENST00000434267.5   MAP4        ATA    158   \n",
       "18823  ENSG00000047849.23   ENST00000439356.2   MAP4        ATG     71   \n",
       "18614  ENSG00000047849.23   ENST00000434267.5   MAP4        ATG     98   \n",
       "\n",
       "         TisType  TISGroup  TISCounts     TISQvalue                 GenomePos  \\\n",
       "16854  Truncated         5         21  1.000000e+00  chr3:47852791-47871239:-   \n",
       "16850  Truncated         5         32  1.000000e+00  chr3:47852791-47872019:-   \n",
       "16853  Truncated         5         21  1.000000e+00  chr3:47852791-47875876:-   \n",
       "16852  Truncated         5         23  1.000000e+00  chr3:47852791-47916119:-   \n",
       "16851  Truncated         5         23  1.000000e+00  chr3:47852791-47916605:-   \n",
       "16849  Annotated         5         91  1.346053e-02  chr3:47852791-47998860:-   \n",
       "24685  Truncated         4         17  1.000000e+00  chr3:47852933-47855319:-   \n",
       "24678  Truncated         4         37  2.341482e-01  chr3:47852933-47867281:-   \n",
       "24687  Truncated         4         13  1.000000e+00  chr3:47852933-47867293:-   \n",
       "24682  Truncated         4         21  1.000000e+00  chr3:47852933-47871239:-   \n",
       "24679  Truncated         4         32  5.004593e-01  chr3:47852933-47872019:-   \n",
       "24683  Truncated         4         21  1.000000e+00  chr3:47852933-47875876:-   \n",
       "24681  Truncated         4         23  1.000000e+00  chr3:47852933-47916119:-   \n",
       "24688  Truncated         4         12  1.000000e+00  chr3:47852933-47916443:-   \n",
       "24680  Truncated         4         23  1.000000e+00  chr3:47852933-47916605:-   \n",
       "24684  Truncated         4         18  1.000000e+00  chr3:47852933-47917043:-   \n",
       "24686  Truncated         4         14  1.000000e+00  chr3:47852933-47917157:-   \n",
       "24677  Annotated         4         91  3.792186e-05  chr3:47852933-47998860:-   \n",
       "11652  Truncated         5         37  1.000000e+00  chr3:47853031-47867281:-   \n",
       "11657  Truncated         5         21  1.000000e+00  chr3:47853031-47871239:-   \n",
       "11653  Truncated         5         32  1.000000e+00  chr3:47853031-47872019:-   \n",
       "11656  Truncated         5         21  1.000000e+00  chr3:47853031-47875876:-   \n",
       "11654  Truncated         5         23  1.000000e+00  chr3:47853031-47916119:-   \n",
       "11655  Truncated         5         23  1.000000e+00  chr3:47853031-47916605:-   \n",
       "11650  Annotated         5         91  1.346053e-02  chr3:47853031-47998860:-   \n",
       "11651   Extended         5         42  1.000000e+00  chr3:47853031-48088870:-   \n",
       "18224  Annotated         6         91  2.265140e-01  chr3:47916863-47998860:-   \n",
       "18615  Truncated         2          8  1.000000e+00  chr3:47975457-47998797:-   \n",
       "18824  Truncated         3          8  1.000000e+00  chr3:47975457-47998797:-   \n",
       "18616  Truncated         2          7  1.000000e+00  chr3:47975457-47998800:-   \n",
       "18823  Annotated         3         91  2.367954e-08  chr3:47975457-47998860:-   \n",
       "18614  Annotated         2         91  1.650892e-12  chr3:47975457-47998860:-   \n",
       "\n",
       "                                                     Seq  \\\n",
       "16854  AAGTCTGTACCAGCTGACTTGAGTCGCCCAAAGAGCACCTCCACCA...   \n",
       "16850  AAGAGCACTCAGACTGTTGCAAAAACCACAACAGCTGCTGCTGTTG...   \n",
       "16853  AAGACTTCAACATCGAAAGCCAAAACACAGCCCACTTCTCTCCCTA...   \n",
       "16852  GTGACTCCAGCCAAAGATGTTCCACCACTCTCAGAAACAGAGGCAA...   \n",
       "16851  AAGGATTTGGTATTACTCTCAGAAATAGAGGTGGCACAGGCTAATG...   \n",
       "16849  ATGGCTGACCTCAGTCTTGCAGATGCATTAACAGAACCATCTCCAG...   \n",
       "24685  AAGGAGAAGGCCCAGGCCAAGGTGGGATCCCTCGATAATGTGGGCC...   \n",
       "24678  AAGGACAATATTAAGCATGTCCCTGGAGGTGGTAATGTTCAGATTC...   \n",
       "24687  AAGTGTGGTTCCAAGGACAATATTAAGCATGTCCCTGGAGGTGGTA...   \n",
       "24682  AAGTCTGTACCAGCTGACTTGAGTCGCCCAAAGAGCACCTCCACCA...   \n",
       "24679  AAGAGCACTCAGACTGTTGCAAAAACCACAACAGCTGCTGCTGTTG...   \n",
       "24683  AAGACTTCAACATCGAAAGCCAAAACACAGCCCACTTCTCTCCCTA...   \n",
       "24681  GTGACTCCAGCCAAAGATGTTCCACCACTCTCAGAAACAGAGGCAA...   \n",
       "24688  AAGGATAAAGCACTACCTTTAGAAGCAGAGGTGGCCCCAGTCAAGG...   \n",
       "24680  AAGGATTTGGTATTACTCTCAGAAATAGAGGTGGCACAGGCTAATG...   \n",
       "24684  AAGGACATGGCACTAGCTACAAAAACCGAGGTGGCATTGGCTAAAG...   \n",
       "24686  AAGGAGATAGAAATGGCATCAGAAGAGAGGCCACCAGCACAAGCAT...   \n",
       "24677  ATGGCTGACCTCAGTCTTGCAGATGCATTAACAGAACCATCTCCAG...   \n",
       "11652  AAGGACAATATTAAGCATGTCCCTGGAGGTGGTAATGTTCAGATTC...   \n",
       "11657  AAGTCTGTACCAGCTGACTTGAGTCGCCCAAAGAGCACCTCCACCA...   \n",
       "11653  AAGAGCACTCAGACTGTTGCAAAAACCACAACAGCTGCTGCTGTTG...   \n",
       "11656  AAGACTTCAACATCGAAAGCCAAAACACAGCCCACTTCTCTCCCTA...   \n",
       "11654  GTGACTCCAGCCAAAGATGTTCCACCACTCTCAGAAACAGAGGCAA...   \n",
       "11655  AAGGATTTGGTATTACTCTCAGAAATAGAGGTGGCACAGGCTAATG...   \n",
       "11650  ATGGCTGACCTCAGTCTTGCAGATGCATTAACAGAACCATCTCCAG...   \n",
       "11651  CTGGCTCCCGGCTCCTCCCGACGTCTCCTACCTCCTCACGGCTCTT...   \n",
       "18224  ATGGCTGACCTCAGTCTTGCAGATGCATTAACAGAACCATCTCCAG...   \n",
       "18615  AAGCGGGACTTCATTGCCACACTAGAGGCAGAGGCCTTTGATGATG...   \n",
       "18824  AAGCGGGACTTCATTGCCACACTAGAGGCAGAGGCCTTTGATGATG...   \n",
       "18616  ATAAAGCGGGACTTCATTGCCACACTAGAGGCAGAGGCCTTTGATG...   \n",
       "18823  ATGGCTGACCTCAGTCTTGCAGATGCATTAACAGAACCATCTCCAG...   \n",
       "18614  ATGGCTGACCTCAGTCTTGCAGATGCATTAACAGAACCATCTCCAG...   \n",
       "\n",
       "                                                   AASeq feature_type strand  \\\n",
       "16854  KSVPADLSRPKSTSTSSMKKTTTLSGTAPAAGVVPSRVKATPMPSR...   transcript      -   \n",
       "16850  KSTQTVAKTTTAAAVASTGPSSRSPSTLLPKKPTAIKTEGKPAEVK...   transcript      -   \n",
       "16853  KTSTSKAKTQPTSLPKQPAPTTIGGLNKKPMSLASGLVPAAPPKRP...   transcript      -   \n",
       "16852  VTPAKDVPPLSETEATPVPIKDMEIAQTQKGISEDSHLESLQDVGQ...   transcript      -   \n",
       "16851  KDLVLLSEIEVAQANDIISSTEISSAEKVALSSETEVALARDMTLP...   transcript      -   \n",
       "16849  MADLSLADALTEPSPDIEGEIKRDFIATLEAEAFDDVVGETVGKTD...   transcript      -   \n",
       "24685  KEKAQAKVGSLDNVGHLPAGGAVKTEGGGSEAPLCPGPPAGEEPAI...   transcript      -   \n",
       "24678  KDNIKHVPGGGNVQIQNKKVDISKVSSKCGSKANIKHKPGGGDVKI...   transcript      -   \n",
       "24687  KCGSKDNIKHVPGGGNVQIQNKKVDISKVSSKCGSKANIKHKPGGG...   transcript      -   \n",
       "24682  KSVPADLSRPKSTSTSSMKKTTTLSGTAPAAGVVPSRVKATPMPSR...   transcript      -   \n",
       "24679  KSTQTVAKTTTAAAVASTGPSSRSPSTLLPKKPTAIKTEGKPAEVK...   transcript      -   \n",
       "24683  KTSTSKAKTQPTSLPKQPAPTTIGGLNKKPMSLASGLVPAAPPKRP...   transcript      -   \n",
       "24681  VTPAKDVPPLSETEATPVPIKDMEIAQTQKGISEDSHLESLQDVGQ...   transcript      -   \n",
       "24688  KDKALPLEAEVAPVKDMAQLPETEIAPAKDVAPSTVKEVGLLKDMS...   transcript      -   \n",
       "24680  KDLVLLSEIEVAQANDIISSTEISSAEKVALSSETEVALARDMTLP...   transcript      -   \n",
       "24684  KDMALATKTEVALAKDMESPTKLDVTLAKDMQPSMESDMALVKDME...   transcript      -   \n",
       "24686  KEIEMASEERPPAQALEIMMGLKTTDMAPSKETEMALAKDMALATK...   transcript      -   \n",
       "24677  MADLSLADALTEPSPDIEGEIKRDFIATLEAEAFDDVVGETVGKTD...   transcript      -   \n",
       "11652  KDNIKHVPGGGNVQIQNKKVDISKVSSKCGSKANIKHKPGGGDVKI...   transcript      -   \n",
       "11657  KSVPADLSRPKSTSTSSMKKTTTLSGTAPAAGVVPSRVKATPMPSR...   transcript      -   \n",
       "11653  KSTQTVAKTTTAAAVASTGPSSRSPSTLLPKKPTAIKTEGKPAEVK...   transcript      -   \n",
       "11656  KTSTSKAKTQPTSLPKQPAPTTIGGLNKKPMSLASGLVPAAPPKRP...   transcript      -   \n",
       "11654  VTPAKDVPPLSETEATPVPIKDMEIAQTQKGISEDSHLESLQDVGQ...   transcript      -   \n",
       "11655  KDLVLLSEIEVAQANDIISSTEISSAEKVALSSETEVALARDMTLP...   transcript      -   \n",
       "11650  MADLSLADALTEPSPDIEGEIKRDFIATLEAEAFDDVVGETVGKTD...   transcript      -   \n",
       "11651  LAPGSSRRLLPPHGSSRRSPGSLLPQLRLGGGGQLQWCRMADLSLA...   transcript      -   \n",
       "18224  MADLSLADALTEPSPDIEGEIKRDFIATLEAEAFDDVVGETVGKTD...   transcript      -   \n",
       "18615  KRDFIATLEAEAFDDVVGETVGKTDYIPLLDVDEKTGNSESKKKPC...   transcript      -   \n",
       "18824  KRDFIATLEAEAFDDVVGETVGKTDYIPLLDVDEKTGNSESKKKPC...   transcript      -   \n",
       "18616  IKRDFIATLEAEAFDDVVGETVGKTDYIPLLDVDEKTGNSESKKKP...   transcript      -   \n",
       "18823  MADLSLADALTEPSPDIEGEIKRDFIATLEAEAFDDVVGETVGKTD...   transcript      -   \n",
       "18614  MADLSLADALTEPSPDIEGEIKRDFIATLEAEAFDDVVGETVGKTD...   transcript      -   \n",
       "\n",
       "      frame       gene_type transcript_type  transcript_support_level  \\\n",
       "16854     .  protein_coding  protein_coding                       2.0   \n",
       "16850     .  protein_coding  protein_coding                       2.0   \n",
       "16853     .  protein_coding  protein_coding                       2.0   \n",
       "16852     .  protein_coding  protein_coding                       2.0   \n",
       "16851     .  protein_coding  protein_coding                       2.0   \n",
       "16849     .  protein_coding  protein_coding                       2.0   \n",
       "24685     .  protein_coding  protein_coding                       NaN   \n",
       "24678     .  protein_coding  protein_coding                       NaN   \n",
       "24687     .  protein_coding  protein_coding                       NaN   \n",
       "24682     .  protein_coding  protein_coding                       NaN   \n",
       "24679     .  protein_coding  protein_coding                       NaN   \n",
       "24683     .  protein_coding  protein_coding                       NaN   \n",
       "24681     .  protein_coding  protein_coding                       NaN   \n",
       "24688     .  protein_coding  protein_coding                       NaN   \n",
       "24680     .  protein_coding  protein_coding                       NaN   \n",
       "24684     .  protein_coding  protein_coding                       NaN   \n",
       "24686     .  protein_coding  protein_coding                       NaN   \n",
       "24677     .  protein_coding  protein_coding                       NaN   \n",
       "11652     .  protein_coding  protein_coding                       1.0   \n",
       "11657     .  protein_coding  protein_coding                       1.0   \n",
       "11653     .  protein_coding  protein_coding                       1.0   \n",
       "11656     .  protein_coding  protein_coding                       1.0   \n",
       "11654     .  protein_coding  protein_coding                       1.0   \n",
       "11655     .  protein_coding  protein_coding                       1.0   \n",
       "11650     .  protein_coding  protein_coding                       1.0   \n",
       "11651     .  protein_coding  protein_coding                       1.0   \n",
       "18224     .  protein_coding  protein_coding                       3.0   \n",
       "18615     .  protein_coding  protein_coding                       2.0   \n",
       "18824     .  protein_coding  protein_coding                       1.0   \n",
       "18616     .  protein_coding  protein_coding                       2.0   \n",
       "18823     .  protein_coding  protein_coding                       1.0   \n",
       "18614     .  protein_coding  protein_coding                       2.0   \n",
       "\n",
       "       MANE_Select  \n",
       "16854        False  \n",
       "16850        False  \n",
       "16853        False  \n",
       "16852        False  \n",
       "16851        False  \n",
       "16849        False  \n",
       "24685         True  \n",
       "24678         True  \n",
       "24687         True  \n",
       "24682         True  \n",
       "24679         True  \n",
       "24683         True  \n",
       "24681         True  \n",
       "24688         True  \n",
       "24680         True  \n",
       "24684         True  \n",
       "24686         True  \n",
       "24677         True  \n",
       "11652        False  \n",
       "11657        False  \n",
       "11653        False  \n",
       "11656        False  \n",
       "11654        False  \n",
       "11655        False  \n",
       "11650        False  \n",
       "11651        False  \n",
       "18224        False  \n",
       "18615        False  \n",
       "18824        False  \n",
       "18616        False  \n",
       "18823        False  \n",
       "18614        False  "
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tis_df[tis_df['Symbol'] == 'MAP4'].sort_values('GenomePos')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e39cc3ce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\nIGV examples: \\nDNM1L => chr12:32,677,451-32,759,074\\nAHNAK => chr11:62,516,743-62,536,098\\nMAP4 => chr3:47853031-48088870\\nASPH => chr8:61503358-61714371\\n'"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "IGV examples: \n",
    "DNM1L => chr12:32677451-32759074 (around start & gene body)\n",
    "AHNAK => chr11:62516743-62536098 (majority of the gene body)\n",
    "MAP4 => chr3:47853031-48088870 (majority of the gene body)\n",
    "ASPH => chr8:61503358-61714371 (around start)\n",
    "\n",
    "Validated real:\n",
    "CDC20\n",
    "AURKAIP1\n",
    "TRMT1\n",
    "UBE2M\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "b4962ccc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gid</th>\n",
       "      <th>Tid</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>StartCodon</th>\n",
       "      <th>Start</th>\n",
       "      <th>TisType</th>\n",
       "      <th>TISGroup</th>\n",
       "      <th>TISCounts</th>\n",
       "      <th>TISQvalue</th>\n",
       "      <th>count_diff</th>\n",
       "      <th>...</th>\n",
       "      <th>GenomePos</th>\n",
       "      <th>Seq</th>\n",
       "      <th>AASeq</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>MANE_Select</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6126</th>\n",
       "      <td>ENSG00000198363.19</td>\n",
       "      <td>ENST00000379454.9</td>\n",
       "      <td>ASPH</td>\n",
       "      <td>AAG</td>\n",
       "      <td>233</td>\n",
       "      <td>Truncated</td>\n",
       "      <td>4</td>\n",
       "      <td>81</td>\n",
       "      <td>1.968915e-04</td>\n",
       "      <td>81</td>\n",
       "      <td>...</td>\n",
       "      <td>chr8:61503358-61714359:-</td>\n",
       "      <td>AAGAATGCCAAGAGCAGCGGCAACAGCAGCAGCAGCGGCTCCGGCA...</td>\n",
       "      <td>KNAKSSGNSSSSGSGSGSTSAGSSSPGARRETKHGGHKNGRKGGLS...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6125</th>\n",
       "      <td>ENSG00000198363.19</td>\n",
       "      <td>ENST00000379454.9</td>\n",
       "      <td>ASPH</td>\n",
       "      <td>ATG</td>\n",
       "      <td>221</td>\n",
       "      <td>Annotated</td>\n",
       "      <td>4</td>\n",
       "      <td>334</td>\n",
       "      <td>8.257402e-23</td>\n",
       "      <td>334</td>\n",
       "      <td>...</td>\n",
       "      <td>chr8:61503358-61714371:-</td>\n",
       "      <td>ATGGCCCAGCGTAAGAATGCCAAGAGCAGCGGCAACAGCAGCAGCA...</td>\n",
       "      <td>MAQRKNAKSSGNSSSSGSGSGSTSAGSSSPGARRETKHGGHKNGRK...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6127</th>\n",
       "      <td>ENSG00000198363.19</td>\n",
       "      <td>ENST00000379454.9</td>\n",
       "      <td>ASPH</td>\n",
       "      <td>CTG</td>\n",
       "      <td>176</td>\n",
       "      <td>Extended</td>\n",
       "      <td>4</td>\n",
       "      <td>151</td>\n",
       "      <td>1.867372e-09</td>\n",
       "      <td>151</td>\n",
       "      <td>...</td>\n",
       "      <td>chr8:61503358-61714416:-</td>\n",
       "      <td>CTGAAGGAGGTCCGCCAGCCCTCACCAGCCCCCGCGGACCGTGCAA...</td>\n",
       "      <td>LKEVRQPSPAPADRAMAQRKNAKSSGNSSSSGSGSGSTSAGSSSPG...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4470</th>\n",
       "      <td>ENSG00000198363.19</td>\n",
       "      <td>ENST00000356457.9</td>\n",
       "      <td>ASPH</td>\n",
       "      <td>ATG</td>\n",
       "      <td>225</td>\n",
       "      <td>Annotated</td>\n",
       "      <td>5</td>\n",
       "      <td>334</td>\n",
       "      <td>6.170442e-13</td>\n",
       "      <td>334</td>\n",
       "      <td>...</td>\n",
       "      <td>chr8:61626272-61714371:-</td>\n",
       "      <td>ATGGCCCAGCGTAAGAATGCCAAGAGCAGCGGCAACAGCAGCAGCA...</td>\n",
       "      <td>MAQRKNAKSSGNSSSSGSGSGSTSAGSSSPGARRETKHGGHKNGRK...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7783</th>\n",
       "      <td>ENSG00000198363.19</td>\n",
       "      <td>ENST00000445642.6</td>\n",
       "      <td>ASPH</td>\n",
       "      <td>ATG</td>\n",
       "      <td>269</td>\n",
       "      <td>Annotated</td>\n",
       "      <td>5</td>\n",
       "      <td>334</td>\n",
       "      <td>6.170442e-13</td>\n",
       "      <td>334</td>\n",
       "      <td>...</td>\n",
       "      <td>chr8:61626272-61714371:-</td>\n",
       "      <td>ATGGCCCAGCGTAAGAATGCCAAGAGCAGCGGCAACAGCAGCAGCA...</td>\n",
       "      <td>MAQRKNAKSSGNSSSSGSGSGSTSAGSSSPGARRETKHGGHKNGRK...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8660</th>\n",
       "      <td>ENSG00000198363.19</td>\n",
       "      <td>ENST00000518068.5</td>\n",
       "      <td>ASPH</td>\n",
       "      <td>ATG</td>\n",
       "      <td>197</td>\n",
       "      <td>Annotated</td>\n",
       "      <td>5</td>\n",
       "      <td>334</td>\n",
       "      <td>6.170442e-13</td>\n",
       "      <td>334</td>\n",
       "      <td>...</td>\n",
       "      <td>chr8:61626272-61714371:-</td>\n",
       "      <td>ATGGCCCAGCGTAAGAATGCCAAGAGCAGCGGCAACAGCAGCAGCA...</td>\n",
       "      <td>MAQRKNAKSSGNSSSSGSGSGSTSAGSSSPGARRETKHGGHKNGRK...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4471</th>\n",
       "      <td>ENSG00000198363.19</td>\n",
       "      <td>ENST00000356457.9</td>\n",
       "      <td>ASPH</td>\n",
       "      <td>CTG</td>\n",
       "      <td>180</td>\n",
       "      <td>Extended</td>\n",
       "      <td>5</td>\n",
       "      <td>151</td>\n",
       "      <td>4.078146e-05</td>\n",
       "      <td>151</td>\n",
       "      <td>...</td>\n",
       "      <td>chr8:61626272-61714416:-</td>\n",
       "      <td>CTGAAGGAGGTCCGCCAGCCCTCACCAGCCCCCGCGGACCGTGCAA...</td>\n",
       "      <td>LKEVRQPSPAPADRAMAQRKNAKSSGNSSSSGSGSGSTSAGSSSPG...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7784</th>\n",
       "      <td>ENSG00000198363.19</td>\n",
       "      <td>ENST00000445642.6</td>\n",
       "      <td>ASPH</td>\n",
       "      <td>CTG</td>\n",
       "      <td>224</td>\n",
       "      <td>Extended</td>\n",
       "      <td>5</td>\n",
       "      <td>151</td>\n",
       "      <td>4.078146e-05</td>\n",
       "      <td>151</td>\n",
       "      <td>...</td>\n",
       "      <td>chr8:61626272-61714416:-</td>\n",
       "      <td>CTGAAGGAGGTCCGCCAGCCCTCACCAGCCCCCGCGGACCGTGCAA...</td>\n",
       "      <td>LKEVRQPSPAPADRAMAQRKNAKSSGNSSSSGSGSGSTSAGSSSPG...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8661</th>\n",
       "      <td>ENSG00000198363.19</td>\n",
       "      <td>ENST00000518068.5</td>\n",
       "      <td>ASPH</td>\n",
       "      <td>CTG</td>\n",
       "      <td>152</td>\n",
       "      <td>Extended</td>\n",
       "      <td>5</td>\n",
       "      <td>151</td>\n",
       "      <td>4.078146e-05</td>\n",
       "      <td>151</td>\n",
       "      <td>...</td>\n",
       "      <td>chr8:61626272-61714416:-</td>\n",
       "      <td>CTGAAGGAGGTCCGCCAGCCCTCACCAGCCCCCGCGGACCGTGCAA...</td>\n",
       "      <td>LKEVRQPSPAPADRAMAQRKNAKSSGNSSSSGSGSGSTSAGSSSPG...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8652</th>\n",
       "      <td>ENSG00000198363.19</td>\n",
       "      <td>ENST00000517856.5</td>\n",
       "      <td>ASPH</td>\n",
       "      <td>AAG</td>\n",
       "      <td>228</td>\n",
       "      <td>Truncated</td>\n",
       "      <td>2</td>\n",
       "      <td>81</td>\n",
       "      <td>5.606415e-11</td>\n",
       "      <td>81</td>\n",
       "      <td>...</td>\n",
       "      <td>chr8:61661906-61714359:-</td>\n",
       "      <td>AAGAATGCCAAGAGCAGCGGCAACAGCAGCAGCAGCGGCTCCGGCA...</td>\n",
       "      <td>KNAKSSGNSSSSGSGSGSTSAGSSSPGARRETKHGGHKNGRKGGLS...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8651</th>\n",
       "      <td>ENSG00000198363.19</td>\n",
       "      <td>ENST00000517856.5</td>\n",
       "      <td>ASPH</td>\n",
       "      <td>ATG</td>\n",
       "      <td>216</td>\n",
       "      <td>Annotated</td>\n",
       "      <td>2</td>\n",
       "      <td>334</td>\n",
       "      <td>3.940734e-50</td>\n",
       "      <td>334</td>\n",
       "      <td>...</td>\n",
       "      <td>chr8:61661906-61714371:-</td>\n",
       "      <td>ATGGCCCAGCGTAAGAATGCCAAGAGCAGCGGCAACAGCAGCAGCA...</td>\n",
       "      <td>MAQRKNAKSSGNSSSSGSGSGSTSAGSSSPGARRETKHGGHKNGRK...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8653</th>\n",
       "      <td>ENSG00000198363.19</td>\n",
       "      <td>ENST00000517856.5</td>\n",
       "      <td>ASPH</td>\n",
       "      <td>CTG</td>\n",
       "      <td>171</td>\n",
       "      <td>Extended</td>\n",
       "      <td>2</td>\n",
       "      <td>151</td>\n",
       "      <td>9.783564e-22</td>\n",
       "      <td>151</td>\n",
       "      <td>...</td>\n",
       "      <td>chr8:61661906-61714416:-</td>\n",
       "      <td>CTGAAGGAGGTCCGCCAGCCCTCACCAGCCCCCGCGGACCGTGCAA...</td>\n",
       "      <td>LKEVRQPSPAPADRAMAQRKNAKSSGNSSSSGSGSGSTSAGSSSPG...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6123</th>\n",
       "      <td>ENSG00000198363.19</td>\n",
       "      <td>ENST00000379449.11</td>\n",
       "      <td>ASPH</td>\n",
       "      <td>ATG</td>\n",
       "      <td>221</td>\n",
       "      <td>Annotated</td>\n",
       "      <td>5</td>\n",
       "      <td>334</td>\n",
       "      <td>6.170442e-13</td>\n",
       "      <td>334</td>\n",
       "      <td>...</td>\n",
       "      <td>chr8:61676074-61714371:-</td>\n",
       "      <td>ATGGCCCAGCGTAAGAATGCCAAGAGCAGCGGCAACAGCAGCAGCA...</td>\n",
       "      <td>MAQRKNAKSSGNSSSSGSGSGSTSAGSSSPGARRETKHGGHKNGRK...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6124</th>\n",
       "      <td>ENSG00000198363.19</td>\n",
       "      <td>ENST00000379449.11</td>\n",
       "      <td>ASPH</td>\n",
       "      <td>CTG</td>\n",
       "      <td>176</td>\n",
       "      <td>Extended</td>\n",
       "      <td>5</td>\n",
       "      <td>151</td>\n",
       "      <td>4.078146e-05</td>\n",
       "      <td>151</td>\n",
       "      <td>...</td>\n",
       "      <td>chr8:61676074-61714416:-</td>\n",
       "      <td>CTGAAGGAGGTCCGCCAGCCCTCACCAGCCCCCGCGGACCGTGCAA...</td>\n",
       "      <td>LKEVRQPSPAPADRAMAQRKNAKSSGNSSSSGSGSGSTSAGSSSPG...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14 rows  21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Gid                 Tid Symbol StartCodon  Start  \\\n",
       "6126  ENSG00000198363.19   ENST00000379454.9   ASPH        AAG    233   \n",
       "6125  ENSG00000198363.19   ENST00000379454.9   ASPH        ATG    221   \n",
       "6127  ENSG00000198363.19   ENST00000379454.9   ASPH        CTG    176   \n",
       "4470  ENSG00000198363.19   ENST00000356457.9   ASPH        ATG    225   \n",
       "7783  ENSG00000198363.19   ENST00000445642.6   ASPH        ATG    269   \n",
       "8660  ENSG00000198363.19   ENST00000518068.5   ASPH        ATG    197   \n",
       "4471  ENSG00000198363.19   ENST00000356457.9   ASPH        CTG    180   \n",
       "7784  ENSG00000198363.19   ENST00000445642.6   ASPH        CTG    224   \n",
       "8661  ENSG00000198363.19   ENST00000518068.5   ASPH        CTG    152   \n",
       "8652  ENSG00000198363.19   ENST00000517856.5   ASPH        AAG    228   \n",
       "8651  ENSG00000198363.19   ENST00000517856.5   ASPH        ATG    216   \n",
       "8653  ENSG00000198363.19   ENST00000517856.5   ASPH        CTG    171   \n",
       "6123  ENSG00000198363.19  ENST00000379449.11   ASPH        ATG    221   \n",
       "6124  ENSG00000198363.19  ENST00000379449.11   ASPH        CTG    176   \n",
       "\n",
       "        TisType  TISGroup  TISCounts     TISQvalue  count_diff  ...  \\\n",
       "6126  Truncated         4         81  1.968915e-04          81  ...   \n",
       "6125  Annotated         4        334  8.257402e-23         334  ...   \n",
       "6127   Extended         4        151  1.867372e-09         151  ...   \n",
       "4470  Annotated         5        334  6.170442e-13         334  ...   \n",
       "7783  Annotated         5        334  6.170442e-13         334  ...   \n",
       "8660  Annotated         5        334  6.170442e-13         334  ...   \n",
       "4471   Extended         5        151  4.078146e-05         151  ...   \n",
       "7784   Extended         5        151  4.078146e-05         151  ...   \n",
       "8661   Extended         5        151  4.078146e-05         151  ...   \n",
       "8652  Truncated         2         81  5.606415e-11          81  ...   \n",
       "8651  Annotated         2        334  3.940734e-50         334  ...   \n",
       "8653   Extended         2        151  9.783564e-22         151  ...   \n",
       "6123  Annotated         5        334  6.170442e-13         334  ...   \n",
       "6124   Extended         5        151  4.078146e-05         151  ...   \n",
       "\n",
       "                     GenomePos  \\\n",
       "6126  chr8:61503358-61714359:-   \n",
       "6125  chr8:61503358-61714371:-   \n",
       "6127  chr8:61503358-61714416:-   \n",
       "4470  chr8:61626272-61714371:-   \n",
       "7783  chr8:61626272-61714371:-   \n",
       "8660  chr8:61626272-61714371:-   \n",
       "4471  chr8:61626272-61714416:-   \n",
       "7784  chr8:61626272-61714416:-   \n",
       "8661  chr8:61626272-61714416:-   \n",
       "8652  chr8:61661906-61714359:-   \n",
       "8651  chr8:61661906-61714371:-   \n",
       "8653  chr8:61661906-61714416:-   \n",
       "6123  chr8:61676074-61714371:-   \n",
       "6124  chr8:61676074-61714416:-   \n",
       "\n",
       "                                                    Seq  \\\n",
       "6126  AAGAATGCCAAGAGCAGCGGCAACAGCAGCAGCAGCGGCTCCGGCA...   \n",
       "6125  ATGGCCCAGCGTAAGAATGCCAAGAGCAGCGGCAACAGCAGCAGCA...   \n",
       "6127  CTGAAGGAGGTCCGCCAGCCCTCACCAGCCCCCGCGGACCGTGCAA...   \n",
       "4470  ATGGCCCAGCGTAAGAATGCCAAGAGCAGCGGCAACAGCAGCAGCA...   \n",
       "7783  ATGGCCCAGCGTAAGAATGCCAAGAGCAGCGGCAACAGCAGCAGCA...   \n",
       "8660  ATGGCCCAGCGTAAGAATGCCAAGAGCAGCGGCAACAGCAGCAGCA...   \n",
       "4471  CTGAAGGAGGTCCGCCAGCCCTCACCAGCCCCCGCGGACCGTGCAA...   \n",
       "7784  CTGAAGGAGGTCCGCCAGCCCTCACCAGCCCCCGCGGACCGTGCAA...   \n",
       "8661  CTGAAGGAGGTCCGCCAGCCCTCACCAGCCCCCGCGGACCGTGCAA...   \n",
       "8652  AAGAATGCCAAGAGCAGCGGCAACAGCAGCAGCAGCGGCTCCGGCA...   \n",
       "8651  ATGGCCCAGCGTAAGAATGCCAAGAGCAGCGGCAACAGCAGCAGCA...   \n",
       "8653  CTGAAGGAGGTCCGCCAGCCCTCACCAGCCCCCGCGGACCGTGCAA...   \n",
       "6123  ATGGCCCAGCGTAAGAATGCCAAGAGCAGCGGCAACAGCAGCAGCA...   \n",
       "6124  CTGAAGGAGGTCCGCCAGCCCTCACCAGCCCCCGCGGACCGTGCAA...   \n",
       "\n",
       "                                                  AASeq feature_type strand  \\\n",
       "6126  KNAKSSGNSSSSGSGSGSTSAGSSSPGARRETKHGGHKNGRKGGLS...   transcript      -   \n",
       "6125  MAQRKNAKSSGNSSSSGSGSGSTSAGSSSPGARRETKHGGHKNGRK...   transcript      -   \n",
       "6127  LKEVRQPSPAPADRAMAQRKNAKSSGNSSSSGSGSGSTSAGSSSPG...   transcript      -   \n",
       "4470  MAQRKNAKSSGNSSSSGSGSGSTSAGSSSPGARRETKHGGHKNGRK...   transcript      -   \n",
       "7783  MAQRKNAKSSGNSSSSGSGSGSTSAGSSSPGARRETKHGGHKNGRK...   transcript      -   \n",
       "8660  MAQRKNAKSSGNSSSSGSGSGSTSAGSSSPGARRETKHGGHKNGRK...   transcript      -   \n",
       "4471  LKEVRQPSPAPADRAMAQRKNAKSSGNSSSSGSGSGSTSAGSSSPG...   transcript      -   \n",
       "7784  LKEVRQPSPAPADRAMAQRKNAKSSGNSSSSGSGSGSTSAGSSSPG...   transcript      -   \n",
       "8661  LKEVRQPSPAPADRAMAQRKNAKSSGNSSSSGSGSGSTSAGSSSPG...   transcript      -   \n",
       "8652  KNAKSSGNSSSSGSGSGSTSAGSSSPGARRETKHGGHKNGRKGGLS...   transcript      -   \n",
       "8651  MAQRKNAKSSGNSSSSGSGSGSTSAGSSSPGARRETKHGGHKNGRK...   transcript      -   \n",
       "8653  LKEVRQPSPAPADRAMAQRKNAKSSGNSSSSGSGSGSTSAGSSSPG...   transcript      -   \n",
       "6123  MAQRKNAKSSGNSSSSGSGSGSTSAGSSSPGARRETKHGGHKNGRK...   transcript      -   \n",
       "6124  LKEVRQPSPAPADRAMAQRKNAKSSGNSSSSGSGSGSTSAGSSSPG...   transcript      -   \n",
       "\n",
       "     frame       gene_type transcript_type transcript_support_level  \\\n",
       "6126     .  protein_coding  protein_coding                      1.0   \n",
       "6125     .  protein_coding  protein_coding                      1.0   \n",
       "6127     .  protein_coding  protein_coding                      1.0   \n",
       "4470     .  protein_coding  protein_coding                      1.0   \n",
       "7783     .  protein_coding  protein_coding                      2.0   \n",
       "8660     .  protein_coding  protein_coding                      1.0   \n",
       "4471     .  protein_coding  protein_coding                      1.0   \n",
       "7784     .  protein_coding  protein_coding                      2.0   \n",
       "8661     .  protein_coding  protein_coding                      1.0   \n",
       "8652     .  protein_coding  protein_coding                      2.0   \n",
       "8651     .  protein_coding  protein_coding                      2.0   \n",
       "8653     .  protein_coding  protein_coding                      2.0   \n",
       "6123     .  protein_coding  protein_coding                      1.0   \n",
       "6124     .  protein_coding  protein_coding                      1.0   \n",
       "\n",
       "      MANE_Select  \n",
       "6126         True  \n",
       "6125         True  \n",
       "6127         True  \n",
       "4470        False  \n",
       "7783        False  \n",
       "8660        False  \n",
       "4471        False  \n",
       "7784        False  \n",
       "8661        False  \n",
       "8652        False  \n",
       "8651        False  \n",
       "8653        False  \n",
       "6123        False  \n",
       "6124        False  \n",
       "\n",
       "[14 rows x 21 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alt_tis_df[alt_tis_df['Gid'] == 'ENSG00000198363.19'].sort_values('GenomePos')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "20b0cb66",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gid</th>\n",
       "      <th>Tid</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>StartCodon</th>\n",
       "      <th>Start</th>\n",
       "      <th>TisType</th>\n",
       "      <th>TISGroup</th>\n",
       "      <th>TISCounts</th>\n",
       "      <th>TISQvalue</th>\n",
       "      <th>GenomePos</th>\n",
       "      <th>Seq</th>\n",
       "      <th>AASeq</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>MANE_Select</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1933</th>\n",
       "      <td>ENSG00000072756.19</td>\n",
       "      <td>ENST00000251607.11</td>\n",
       "      <td>TRNT1</td>\n",
       "      <td>ATG</td>\n",
       "      <td>78</td>\n",
       "      <td>Annotated</td>\n",
       "      <td>1</td>\n",
       "      <td>20</td>\n",
       "      <td>0.003329</td>\n",
       "      <td>chr3:3129040-3148154:+</td>\n",
       "      <td>ATGCTGAGGTGCCTGTATCATTGGCACAGGCCAGTGCTGAACCGTA...</td>\n",
       "      <td>MLRCLYHWHRPVLNRRWSRLCLPKQYLFTMKLQSPEFQSLFTEGLK...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1934</th>\n",
       "      <td>ENSG00000072756.19</td>\n",
       "      <td>ENST00000251607.11</td>\n",
       "      <td>TRNT1</td>\n",
       "      <td>ATG</td>\n",
       "      <td>165</td>\n",
       "      <td>Truncated</td>\n",
       "      <td>1</td>\n",
       "      <td>20</td>\n",
       "      <td>0.003329</td>\n",
       "      <td>chr3:3129127-3148154:+</td>\n",
       "      <td>ATGAAGTTGCAGTCTCCCGAATTCCAGTCACTTTTCACAGAAGGAC...</td>\n",
       "      <td>MKLQSPEFQSLFTEGLKSLTELFVKENHELRIAGGAVRDLLNGVKP...</td>\n",
       "      <td>transcript</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Gid                 Tid Symbol StartCodon  Start  \\\n",
       "1933  ENSG00000072756.19  ENST00000251607.11  TRNT1        ATG     78   \n",
       "1934  ENSG00000072756.19  ENST00000251607.11  TRNT1        ATG    165   \n",
       "\n",
       "        TisType  TISGroup  TISCounts  TISQvalue               GenomePos  \\\n",
       "1933  Annotated         1         20   0.003329  chr3:3129040-3148154:+   \n",
       "1934  Truncated         1         20   0.003329  chr3:3129127-3148154:+   \n",
       "\n",
       "                                                    Seq  \\\n",
       "1933  ATGCTGAGGTGCCTGTATCATTGGCACAGGCCAGTGCTGAACCGTA...   \n",
       "1934  ATGAAGTTGCAGTCTCCCGAATTCCAGTCACTTTTCACAGAAGGAC...   \n",
       "\n",
       "                                                  AASeq feature_type strand  \\\n",
       "1933  MLRCLYHWHRPVLNRRWSRLCLPKQYLFTMKLQSPEFQSLFTEGLK...   transcript      +   \n",
       "1934  MKLQSPEFQSLFTEGLKSLTELFVKENHELRIAGGAVRDLLNGVKP...   transcript      +   \n",
       "\n",
       "     frame       gene_type transcript_type  transcript_support_level  \\\n",
       "1933     .  protein_coding  protein_coding                       1.0   \n",
       "1934     .  protein_coding  protein_coding                       1.0   \n",
       "\n",
       "      MANE_Select  \n",
       "1933         True  \n",
       "1934         True  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tis_df[tis_df['Symbol'] == 'TRNT1'].sort_values('GenomePos')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8d3d8f5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# standard data analysis\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "import os\n",
    "\n",
    "# convenience\n",
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "9c7dc77b",
   "metadata": {},
   "outputs": [],
   "source": [
    "replicate_df = pd.read_csv('ribotish_replicate_manifest.csv')\n",
    "sample_df = pd.read_csv('ribotish_sample_manifest.csv')\n",
    "experiment_table = sample_df.merge(\n",
    "    replicate_df[\n",
    "        replicate_df['condition'] == 'TIS'\n",
    "    ].groupby('sample').apply(lambda x: list(x['bam_qc_file'])).rename('bam_qc_file'), left_on='sample', right_index=True\n",
    ").set_index('sample')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "22ac8ea2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>replicate</th>\n",
       "      <th>condition</th>\n",
       "      <th>bam_qc_file</th>\n",
       "      <th>offset_file</th>\n",
       "      <th>predict_file</th>\n",
       "      <th>filtered_file</th>\n",
       "      <th>dropped_file</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HeLa</td>\n",
       "      <td>rep1</td>\n",
       "      <td>TIS</td>\n",
       "      <td>/lab/cheeseman_sequencing/Jimmy_HTPdata/Riboso...</td>\n",
       "      <td>/lab/barcheese01/aTIS_data/ribosome_profiling/...</td>\n",
       "      <td>/lab/barcheese01/aTIS_data/ribotish/per_rep/He...</td>\n",
       "      <td>/lab/barcheese01/smaffa/filtered_tis_data/per_...</td>\n",
       "      <td>/lab/barcheese01/smaffa/filtered_tis_data/per_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HeLa</td>\n",
       "      <td>rep2</td>\n",
       "      <td>TIS</td>\n",
       "      <td>/lab/cheeseman_sequencing/Jimmy_HTPdata/Riboso...</td>\n",
       "      <td>/lab/barcheese01/aTIS_data/ribosome_profiling/...</td>\n",
       "      <td>/lab/barcheese01/aTIS_data/ribotish/per_rep/He...</td>\n",
       "      <td>/lab/barcheese01/smaffa/filtered_tis_data/per_...</td>\n",
       "      <td>/lab/barcheese01/smaffa/filtered_tis_data/per_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>K562</td>\n",
       "      <td>rep1</td>\n",
       "      <td>TIS</td>\n",
       "      <td>/lab/cheeseman_sequencing/Jimmy_HTPdata/Riboso...</td>\n",
       "      <td>/lab/barcheese01/aTIS_data/ribosome_profiling/...</td>\n",
       "      <td>/lab/barcheese01/aTIS_data/ribotish/per_rep/K5...</td>\n",
       "      <td>/lab/barcheese01/smaffa/filtered_tis_data/per_...</td>\n",
       "      <td>/lab/barcheese01/smaffa/filtered_tis_data/per_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>K562</td>\n",
       "      <td>rep2</td>\n",
       "      <td>TIS</td>\n",
       "      <td>/lab/cheeseman_sequencing/Jimmy_HTPdata/Riboso...</td>\n",
       "      <td>/lab/barcheese01/aTIS_data/ribosome_profiling/...</td>\n",
       "      <td>/lab/barcheese01/aTIS_data/ribotish/per_rep/K5...</td>\n",
       "      <td>/lab/barcheese01/smaffa/filtered_tis_data/per_...</td>\n",
       "      <td>/lab/barcheese01/smaffa/filtered_tis_data/per_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>RPE1_Async</td>\n",
       "      <td>rep1</td>\n",
       "      <td>TIS</td>\n",
       "      <td>/lab/cheeseman_sequencing/Jimmy_HTPdata/Riboso...</td>\n",
       "      <td>/lab/barcheese01/aTIS_data/ribosome_profiling/...</td>\n",
       "      <td>/lab/barcheese01/aTIS_data/ribotish/per_rep/RP...</td>\n",
       "      <td>/lab/barcheese01/smaffa/filtered_tis_data/per_...</td>\n",
       "      <td>/lab/barcheese01/smaffa/filtered_tis_data/per_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>RPE1_Async</td>\n",
       "      <td>rep2</td>\n",
       "      <td>TIS</td>\n",
       "      <td>/lab/cheeseman_sequencing/Jimmy_HTPdata/Riboso...</td>\n",
       "      <td>/lab/barcheese01/aTIS_data/ribosome_profiling/...</td>\n",
       "      <td>/lab/barcheese01/aTIS_data/ribotish/per_rep/RP...</td>\n",
       "      <td>/lab/barcheese01/smaffa/filtered_tis_data/per_...</td>\n",
       "      <td>/lab/barcheese01/smaffa/filtered_tis_data/per_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>RPE1_Que</td>\n",
       "      <td>rep1</td>\n",
       "      <td>TIS</td>\n",
       "      <td>/lab/cheeseman_sequencing/Jimmy_HTPdata/Riboso...</td>\n",
       "      <td>/lab/barcheese01/aTIS_data/ribosome_profiling/...</td>\n",
       "      <td>/lab/barcheese01/aTIS_data/ribotish/per_rep/RP...</td>\n",
       "      <td>/lab/barcheese01/smaffa/filtered_tis_data/per_...</td>\n",
       "      <td>/lab/barcheese01/smaffa/filtered_tis_data/per_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>RPE1_Que</td>\n",
       "      <td>rep2</td>\n",
       "      <td>TIS</td>\n",
       "      <td>/lab/cheeseman_sequencing/Jimmy_HTPdata/Riboso...</td>\n",
       "      <td>/lab/barcheese01/aTIS_data/ribosome_profiling/...</td>\n",
       "      <td>/lab/barcheese01/aTIS_data/ribotish/per_rep/RP...</td>\n",
       "      <td>/lab/barcheese01/smaffa/filtered_tis_data/per_...</td>\n",
       "      <td>/lab/barcheese01/smaffa/filtered_tis_data/per_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>RPE1_Sen</td>\n",
       "      <td>rep1</td>\n",
       "      <td>TIS</td>\n",
       "      <td>/lab/cheeseman_sequencing/Jimmy_HTPdata/Riboso...</td>\n",
       "      <td>/lab/barcheese01/aTIS_data/ribosome_profiling/...</td>\n",
       "      <td>/lab/barcheese01/aTIS_data/ribotish/per_rep/RP...</td>\n",
       "      <td>/lab/barcheese01/smaffa/filtered_tis_data/per_...</td>\n",
       "      <td>/lab/barcheese01/smaffa/filtered_tis_data/per_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>RPE1_Sen</td>\n",
       "      <td>rep2</td>\n",
       "      <td>TIS</td>\n",
       "      <td>/lab/cheeseman_sequencing/Jimmy_HTPdata/Riboso...</td>\n",
       "      <td>/lab/barcheese01/aTIS_data/ribosome_profiling/...</td>\n",
       "      <td>/lab/barcheese01/aTIS_data/ribotish/per_rep/RP...</td>\n",
       "      <td>/lab/barcheese01/smaffa/filtered_tis_data/per_...</td>\n",
       "      <td>/lab/barcheese01/smaffa/filtered_tis_data/per_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>U2OS</td>\n",
       "      <td>rep1</td>\n",
       "      <td>TIS</td>\n",
       "      <td>/lab/cheeseman_sequencing/Jimmy_HTPdata/Riboso...</td>\n",
       "      <td>/lab/barcheese01/aTIS_data/ribosome_profiling/...</td>\n",
       "      <td>/lab/barcheese01/aTIS_data/ribotish/per_rep/U2...</td>\n",
       "      <td>/lab/barcheese01/smaffa/filtered_tis_data/per_...</td>\n",
       "      <td>/lab/barcheese01/smaffa/filtered_tis_data/per_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>U2OS</td>\n",
       "      <td>rep2</td>\n",
       "      <td>TIS</td>\n",
       "      <td>/lab/cheeseman_sequencing/Jimmy_HTPdata/Riboso...</td>\n",
       "      <td>/lab/barcheese01/aTIS_data/ribosome_profiling/...</td>\n",
       "      <td>/lab/barcheese01/aTIS_data/ribotish/per_rep/U2...</td>\n",
       "      <td>/lab/barcheese01/smaffa/filtered_tis_data/per_...</td>\n",
       "      <td>/lab/barcheese01/smaffa/filtered_tis_data/per_...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        sample replicate condition  \\\n",
       "1         HeLa      rep1       TIS   \n",
       "3         HeLa      rep2       TIS   \n",
       "5         K562      rep1       TIS   \n",
       "7         K562      rep2       TIS   \n",
       "9   RPE1_Async      rep1       TIS   \n",
       "11  RPE1_Async      rep2       TIS   \n",
       "13    RPE1_Que      rep1       TIS   \n",
       "15    RPE1_Que      rep2       TIS   \n",
       "17    RPE1_Sen      rep1       TIS   \n",
       "19    RPE1_Sen      rep2       TIS   \n",
       "21        U2OS      rep1       TIS   \n",
       "23        U2OS      rep2       TIS   \n",
       "\n",
       "                                          bam_qc_file  \\\n",
       "1   /lab/cheeseman_sequencing/Jimmy_HTPdata/Riboso...   \n",
       "3   /lab/cheeseman_sequencing/Jimmy_HTPdata/Riboso...   \n",
       "5   /lab/cheeseman_sequencing/Jimmy_HTPdata/Riboso...   \n",
       "7   /lab/cheeseman_sequencing/Jimmy_HTPdata/Riboso...   \n",
       "9   /lab/cheeseman_sequencing/Jimmy_HTPdata/Riboso...   \n",
       "11  /lab/cheeseman_sequencing/Jimmy_HTPdata/Riboso...   \n",
       "13  /lab/cheeseman_sequencing/Jimmy_HTPdata/Riboso...   \n",
       "15  /lab/cheeseman_sequencing/Jimmy_HTPdata/Riboso...   \n",
       "17  /lab/cheeseman_sequencing/Jimmy_HTPdata/Riboso...   \n",
       "19  /lab/cheeseman_sequencing/Jimmy_HTPdata/Riboso...   \n",
       "21  /lab/cheeseman_sequencing/Jimmy_HTPdata/Riboso...   \n",
       "23  /lab/cheeseman_sequencing/Jimmy_HTPdata/Riboso...   \n",
       "\n",
       "                                          offset_file  \\\n",
       "1   /lab/barcheese01/aTIS_data/ribosome_profiling/...   \n",
       "3   /lab/barcheese01/aTIS_data/ribosome_profiling/...   \n",
       "5   /lab/barcheese01/aTIS_data/ribosome_profiling/...   \n",
       "7   /lab/barcheese01/aTIS_data/ribosome_profiling/...   \n",
       "9   /lab/barcheese01/aTIS_data/ribosome_profiling/...   \n",
       "11  /lab/barcheese01/aTIS_data/ribosome_profiling/...   \n",
       "13  /lab/barcheese01/aTIS_data/ribosome_profiling/...   \n",
       "15  /lab/barcheese01/aTIS_data/ribosome_profiling/...   \n",
       "17  /lab/barcheese01/aTIS_data/ribosome_profiling/...   \n",
       "19  /lab/barcheese01/aTIS_data/ribosome_profiling/...   \n",
       "21  /lab/barcheese01/aTIS_data/ribosome_profiling/...   \n",
       "23  /lab/barcheese01/aTIS_data/ribosome_profiling/...   \n",
       "\n",
       "                                         predict_file  \\\n",
       "1   /lab/barcheese01/aTIS_data/ribotish/per_rep/He...   \n",
       "3   /lab/barcheese01/aTIS_data/ribotish/per_rep/He...   \n",
       "5   /lab/barcheese01/aTIS_data/ribotish/per_rep/K5...   \n",
       "7   /lab/barcheese01/aTIS_data/ribotish/per_rep/K5...   \n",
       "9   /lab/barcheese01/aTIS_data/ribotish/per_rep/RP...   \n",
       "11  /lab/barcheese01/aTIS_data/ribotish/per_rep/RP...   \n",
       "13  /lab/barcheese01/aTIS_data/ribotish/per_rep/RP...   \n",
       "15  /lab/barcheese01/aTIS_data/ribotish/per_rep/RP...   \n",
       "17  /lab/barcheese01/aTIS_data/ribotish/per_rep/RP...   \n",
       "19  /lab/barcheese01/aTIS_data/ribotish/per_rep/RP...   \n",
       "21  /lab/barcheese01/aTIS_data/ribotish/per_rep/U2...   \n",
       "23  /lab/barcheese01/aTIS_data/ribotish/per_rep/U2...   \n",
       "\n",
       "                                        filtered_file  \\\n",
       "1   /lab/barcheese01/smaffa/filtered_tis_data/per_...   \n",
       "3   /lab/barcheese01/smaffa/filtered_tis_data/per_...   \n",
       "5   /lab/barcheese01/smaffa/filtered_tis_data/per_...   \n",
       "7   /lab/barcheese01/smaffa/filtered_tis_data/per_...   \n",
       "9   /lab/barcheese01/smaffa/filtered_tis_data/per_...   \n",
       "11  /lab/barcheese01/smaffa/filtered_tis_data/per_...   \n",
       "13  /lab/barcheese01/smaffa/filtered_tis_data/per_...   \n",
       "15  /lab/barcheese01/smaffa/filtered_tis_data/per_...   \n",
       "17  /lab/barcheese01/smaffa/filtered_tis_data/per_...   \n",
       "19  /lab/barcheese01/smaffa/filtered_tis_data/per_...   \n",
       "21  /lab/barcheese01/smaffa/filtered_tis_data/per_...   \n",
       "23  /lab/barcheese01/smaffa/filtered_tis_data/per_...   \n",
       "\n",
       "                                         dropped_file  \n",
       "1   /lab/barcheese01/smaffa/filtered_tis_data/per_...  \n",
       "3   /lab/barcheese01/smaffa/filtered_tis_data/per_...  \n",
       "5   /lab/barcheese01/smaffa/filtered_tis_data/per_...  \n",
       "7   /lab/barcheese01/smaffa/filtered_tis_data/per_...  \n",
       "9   /lab/barcheese01/smaffa/filtered_tis_data/per_...  \n",
       "11  /lab/barcheese01/smaffa/filtered_tis_data/per_...  \n",
       "13  /lab/barcheese01/smaffa/filtered_tis_data/per_...  \n",
       "15  /lab/barcheese01/smaffa/filtered_tis_data/per_...  \n",
       "17  /lab/barcheese01/smaffa/filtered_tis_data/per_...  \n",
       "19  /lab/barcheese01/smaffa/filtered_tis_data/per_...  \n",
       "21  /lab/barcheese01/smaffa/filtered_tis_data/per_...  \n",
       "23  /lab/barcheese01/smaffa/filtered_tis_data/per_...  "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "replicate_df.dropna(subset=['predict_file'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "3a3baa3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_to_filtered_tis = dict()\n",
    "for sample in experiment_table.index.tolist():\n",
    "    pred_table = pd.read_csv(experiment_table.loc[sample, 'predict_file'], sep='\\t', usecols=['Tid', 'GenomePos', 'Start', 'AASeq'])\n",
    "    filt_table = pd.read_csv(experiment_table.loc[sample, 'filtered_file'])\n",
    "    sample_to_filtered_tis[sample] = filt_table.merge(pred_table, left_on=['Tid', 'GenomePos', 'Start'], right_on=['Tid', 'GenomePos', 'Start'], how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "9c5ef221",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_tis_df= pd.concat(sample_to_filtered_tis, axis=0).reset_index(names=['Sample', 'drop']).drop(['drop'], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "672b1bc5",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_tis_df.to_csv('/lab/barcheese01/smaffa/filtered_tis_data/all_samples_filtered.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "1821353e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "sample\n",
       "HeLa          /lab/barcheese01/smaffa/latest_filtered_tis_da...\n",
       "K562          /lab/barcheese01/smaffa/latest_filtered_tis_da...\n",
       "RPE1_Async    /lab/barcheese01/smaffa/latest_filtered_tis_da...\n",
       "RPE1_Que      /lab/barcheese01/smaffa/latest_filtered_tis_da...\n",
       "RPE1_Sen      /lab/barcheese01/smaffa/latest_filtered_tis_da...\n",
       "U2OS          /lab/barcheese01/smaffa/latest_filtered_tis_da...\n",
       "Name: dropped_file, dtype: str"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "experiment_table['dropped_file']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d45d369d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "riboseq",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
